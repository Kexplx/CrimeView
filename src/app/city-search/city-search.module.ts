import { NgModule } from '@angular/core';
import { CityFormComponent } from './city-form/city-form.component';
import { SharedModule } from '../shared/shared.module';
import { CityInputComponent } from './city-input/city-input.component';
import { PredictionPipe } from './city-input/prediction.pipe';

@NgModule({
  declarations: [
    CityInputComponent,
    CityFormComponent,
    PredictionPipe,
  ],
  imports: [SharedModule],
  providers: [google.maps.Geocoder, google.maps.places.AutocompleteService],
  exports: [CityFormComponent],
})
export class CitySearchModule {}
