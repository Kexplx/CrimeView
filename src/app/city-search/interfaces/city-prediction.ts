/**
 * Represents the prediction of a city-search, based on previous user input.
 *
 * "Reg" -> {placeId: "0129", name: "Regensburg"};
 */
export interface CityPrediction {
  placeId: string;
  name: string;
}
