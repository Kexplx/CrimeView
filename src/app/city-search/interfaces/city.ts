export interface Position {
  lat: number;
  lng: number;
}

export interface City {
  placeId: string;
  name: string;
  position: Position;
}
