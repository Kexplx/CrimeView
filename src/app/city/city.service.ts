import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface CityPrediction {
  placeId: string;
  name: string;
}

export interface City {
  placeId: string;
  name: string;
  state: string;
  state_short: string;
  lat: number;
  lng: number;
}

@Injectable()
export class CityService {
  constructor(
    private predictionsService: google.maps.places.AutocompleteService,
    private geoCoder: google.maps.Geocoder,
  ) {}

  private predictionsOptions = {
    componentRestrictions: { country: 'de' },
    types: ['(cities)'],
  };

  getCityPredictions(input: string): Observable<CityPrediction[]> {
    return new Observable(sub => {
      this.predictionsService.getPlacePredictions(
        { ...this.predictionsOptions, input },
        predictions => {
          const cityPredictions = predictions.map<CityPrediction>(p => ({
            name: p.description,
            placeId: p.place_id,
          }));

          sub.next(cityPredictions);
        },
      );
    });
  }

  getCity(placeId: string): Observable<City> {
    return new Observable(sub => {
      this.geoCoder.geocode({ placeId }, result => {
        const { address_components, geometry } = result[0];

        const city: City = {
          placeId,
          lat: geometry.location.lat(),
          lng: geometry.location.lng(),
          name: address_components[0].long_name,
          state: address_components[3].long_name,
          state_short: address_components[3].short_name,
        };

        sub.next(city);
      });
    });
  }
}
