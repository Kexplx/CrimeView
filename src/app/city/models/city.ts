export interface City {
  placeId: string;
  name: string;
  lat: number;
  lng: number;
}
