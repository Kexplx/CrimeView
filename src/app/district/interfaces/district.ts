import { Geometry } from 'geojson';

export type DistrictType =
  | 'Kreis'
  | 'Kreisfreie Stadt'
  | 'Landkreis'
  | 'Regionalverbund'
  | 'Stadtkreis';

/**
 * Represents a german district.
 */
export interface District {
  code: string; // A unique identifier with 5 characters.
  name: string;
  type: DistrictType;
  crestUrl: string; // The url of the districts's crest as svg.
  state: string; // The name of the federal state containing the district.

  totalOffencesCount: number;
  relativeOffencesCount: number; // Number of offences per 100.000 inhabitants.
  commonOffences: { offenceName: string; offencesCount: number }[];
  totalSuspectsCount: number;
  nonGermanSuspectsCount: number;
  solvedCasesCount: number;

  geometry?: Geometry;
}
