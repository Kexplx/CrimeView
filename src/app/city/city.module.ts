import { NgModule } from '@angular/core';
import { CityListComponent } from './city-list/city-list.component';
import { SharedModule } from '../shared/shared.module';
import { CityService } from './city.service';
import { CityInputComponent } from './city-input/city-input.component';

@NgModule({
  declarations: [CityInputComponent, CityListComponent],
  imports: [SharedModule],
  providers: [
    CityService,
    google.maps.Geocoder,
    google.maps.places.AutocompleteService,
  ],
  exports: [CityListComponent],
})
export class CityModule {}
