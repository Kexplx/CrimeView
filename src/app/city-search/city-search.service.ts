import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { DistrictService } from '../district/district.service';
import { District } from '../district/interfaces/district';
import { City } from './interfaces/city';

export interface Search {
  type: 'Polygon' | 'Line' | 'Radius';
  cities: City[];
  districts: District[];
}

@Injectable({
  providedIn: 'root',
})
export class CitySearchService {
  private searchSubject = new Subject<Search>();
  search$ = this.searchSubject.asObservable();

  constructor(private districtService: DistrictService) {}

  handleSearchRequest(cities: City[]): void {
    this.districtService.getDistricts(cities).subscribe(districts => {
      const route: Search = {
        cities,
        districts,
        type: this.getSearchType(cities),
      };

      this.searchSubject.next(route);
    });
  }

  private getSearchType({ length }: City[]): 'Polygon' | 'Line' | 'Radius' {
    return length === 1 ? 'Radius' : length === 2 ? 'Line' : 'Polygon';
  }
}
