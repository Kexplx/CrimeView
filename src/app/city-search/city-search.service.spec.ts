import { District } from '../district/interfaces/district';
import { DistrictService } from '../district/district.service';
import { City } from './interfaces/city';
import { of } from 'rxjs';
import { CitySearchService, Search } from './city-search.service';

let searchService: CitySearchService;

let districtServiceSpy: { getDistricts: jest.Mock };
let dummyDistricts: District[];
let dummyCities: City[];

beforeEach(() => {
  dummyDistricts = [{ name: 'Dummy1' } as District, { name: 'Dummy2' } as District];
  dummyCities = [{ name: 'Dummy3' } as City, { name: 'Dummy4' } as City];
  districtServiceSpy = { getDistricts: jest.fn(() => of(dummyDistricts)) };

  searchService = new CitySearchService((districtServiceSpy as unknown) as DistrictService);
});

describe('#handleSearchRequest', () => {
  it('should call #getDistricts', () => {
    searchService.handleSearchRequest(dummyCities);
    expect(districtServiceSpy.getDistricts).toHaveBeenCalled();
  });

  it('search$ should emit the new search', done => {
    searchService.search$.subscribe(search => {
      expect(search).toEqual<Search>({
        districts: dummyDistricts,
        cities: dummyCities,
        type: 'Line',
      });
      done();
    });

    searchService.handleSearchRequest(dummyCities);
  });
});
