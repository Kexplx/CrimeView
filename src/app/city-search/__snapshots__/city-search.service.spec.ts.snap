// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#handleSearchRequest search$ should emit the new search 1`] = `
Object {
  "cities": Array [
    Object {
      "name": "Dummy3",
    },
    Object {
      "name": "Dummy4",
    },
  ],
  "districts": Array [
    Object {
      "name": "Dummy1",
    },
    Object {
      "name": "Dummy2",
    },
  ],
  "type": "Line",
}
`;
